jdaviz/core/template_mixin.py

```diff
@@ -225,20 +228,56 @@ class PluginTemplateMixin(TemplateMixin):
     This base class can be inherited by all sidebar/tray plugins to expose common functionality.
     """
     disabled_msg = Unicode("").tag(sync=True)
-    plugin_opened = Bool(False).tag(sync=True)
+    plugin_opened = Bool(False).tag(sync=True)  # noqa any instance of the plugin is open (recently sent an "alive" ping)
+    uses_active_status = Bool(False).tag(sync=True)  # noqa whether the plugin has live-preview marks, set to True in plugins to expose keep_active switch
+    keep_active = Bool(False).tag(sync=True)  # noqa whether the live-preview marks show regardless of active state, inapplicable unless uses_active_status is True
+    is_active = Bool(False).tag(sync=True)  # noqa read-only: whether the previews should be shown according to plugin_opened and keep_active
 
     def __init__(self, **kwargs):
         self._viewer_callbacks = {}
+        self._inactive_thread = None  # thread checking for alive pings to control plugin_opened
+        self._ping_timestamp = 0
         super().__init__(**kwargs)
-        self.app.state.add_callback('tray_items_open', self._mxn_update_plugin_opened)
-        self.app.state.add_callback('drawer', self._mxn_update_plugin_opened)
 
     @property
     def user_api(self):
         # plugins should override this to pass their own list of expose functionality, which
         # can even be dependent on config, etc.
         return PluginUserApi(self, expose=[])
 
+    def vue_plugin_ping(self, ping_timestamp):
+        self._ping_timestamp = ping_timestamp
+
+        # we've received a ping, so immediately set plugin_opened state to True
+        if not self.plugin_opened:
+            self.plugin_opened = True
+
+        if self._inactive_thread is not None and self._inactive_thread.is_alive():
+            # a thread already exists to check for pings, the latest ping will allow
+            # the existing while-loop to continue
+            return
+
+        # create a thread to monitor for pings.  If a ping hasn't been received in the
+        # expected time, then plugin_opened will be set to False.
+        self._inactive_thread = threading.Thread(target=self._watch_active)
+        self._inactive_thread.start()
+
+    def _watch_active(self):
+        # expected_delay_ms hould match value in setTimeout in tray_plugin.vue
+        # NOTE: could control with a traitlet, but then would need to pass through each
+        # <j-tray-plugin> component
+        expected_delay_ms = 200
+        # plugin_ping (ms) set by setTimeout in tray_plugin.vue
+        # time.time() is in s, so need to convert to ms
+        while time.time()*1000 - self._ping_timestamp < 2 * expected_delay_ms:
```

Comment:
Because this way `expected_delay_ms` matches the value in javascript (as the comments state).  We could have a separate variable that converts to seconds and adds the multiplicative factor, but as we tweak the delays, we might want different factors in the for loop and the sleep 🤷‍♂️

Reply:
But right now they are the same, and seems like a lot of repeated calculations that are unnecessary. 🤷